include/EigenDecomposition.h

```diff
@@ -290,21 +298,97 @@ void reconstruct_matrix_from_decomposition(const T (&D)[3][3],
 //------------------ unsym_matrix_force_sym_3D -----------------------------
 //--------------------------------------------------------------------------
 template <class T>
-KOKKOS_FUNCTION
-void unsym_matrix_force_sym(T (&A)[3][3], T (&Q)[3][3], T (&D)[3][3]) {
+KOKKOS_FUNCTION void
+  unsym_matrix_force_sym(T (&A)[3][3], T (&Q)[3][3], T (&D)[3][3])
+{
 
-  // force symmetry force 
+  // force symmetry force
   for (int i = 0; i < 3; i++) {
     for (int j = i; j < 3; j++) {
-      A[i][j] = (A[i][j] + A[j][i])/2.0;
+      A[i][j] = (A[i][j] + A[j][i]) / 2.0;
       A[j][i] = A[i][j];
     }
   }
 
-  // then call symmetric diagonalize 
+  // then call symmetric diagonalize
   sym_diagonalize(A, Q, D);
 }
 
+//--------------------------------------------------------------------------
+//------------------ general_eigenvalues_3D -----------------------------
+//--------------------------------------------------------------------------
+template <class T>
+KOKKOS_FUNCTION void
+  general_eigenvalues(T (&A)[3][3], T (&Q)[3][3], T (&D)[3][3])
+{
+
+  const T pi = stk::math::acos(-1.0);
+
+  // Characteristic equation for A is ax^3 + bx^2 + cx + d = 0 where x are the
+  // eigenvalues and a = 1, b = -trA, c = coFacA, d = -detA
+  const T trA = A[0][0] + A[1][1] + A[2][2];
+  const T detA = A[0][0] * A[1][1] * A[2][2] + A[0][1] * A[1][2] * A[2][0] +
+                 A[0][2] * A[1][0] * A[2][1] - A[0][0] * A[1][2] * A[2][1] -
+                 A[0][1] * A[1][0] * A[2][2] - A[0][2] * A[1][1] * A[2][0];
+  const T coFacA = A[0][0] * A[1][1] - A[0][1] * A[1][0] + A[1][1] * A[2][2] -
+                   A[1][2] * A[2][1] + A[0][0] * A[2][2] - A[0][2] * A[2][0];
+
+  // Check to make sure all eigenvalues are real
+  // discriminant = (bc)^2 - 4ac^3 -4b^3d -27a^2d^2 + 18abcd where
+  // a = 1, b = -trA, c = coFacA, d = -detA
+  const T disc = trA * trA * coFacA * coFacA - 4.0 * coFacA * coFacA * coFacA -
+                 4.0 * trA * trA * trA * detA - 27.0 * detA * detA +
+                 18.0 * trA * coFacA * detA;
+
+  const auto check_one = disc < 0.0;
+  const bool exit_now = stk::simd::are_all(check_one);
+  if (exit_now) {
+    NaluEnv::self().naluOutputP0()
+      << "Error, complex eigenvalues in EigenDecomposition::general_eigenvalues"
+      << disc << "([[" << A[0][0] << "," << A[0][1] << "," << A[0][2] << "],["
+      << A[1][0] << "," << A[1][1] << "," << A[1][2] << "],[" << A[2][0] << ","
+      << A[2][1] << "," << A[2][2] << "]])" << std::endl;
+    throw std::runtime_error("Exiting... ");
+  }
```

Comment:
In MPI runs, the `exit_now` situation might occur on an mpi rank that is not the root process. In this situation, the code will exit with an obscure "Exiting..." message while the error details are sent to `/dev/null` by `naluOutputP0`. In general I also recommend repeating the general error message 
`Error, complex eigenvalues in EigenDecomposition::general_eigenvalues` in the runtime_error argument also. 

```suggestion
  if (exit_now) {
    NaluEnv::self().naluOutput()
      << "Error, complex eigenvalues in EigenDecomposition::general_eigenvalues"
      << disc << "([[" << A[0][0] << "," << A[0][1] << "," << A[0][2] << "],["
      << A[1][0] << "," << A[1][1] << "," << A[1][2] << "],[" << A[2][0] << ","
      << A[2][1] << "," << A[2][2] << "]])" << std::endl;
    throw std::runtime_error("Exiting... ");
  }
```

Reply:
Thanks, updated!